sql/distsqlrun: TestIndexSkipTableReader failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/76d0d3ce063db124fe8b4a91c4964c220491a9b5

Parameters:
```
TAGS=
GOFLAGS=-race -parallel=2
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestIndexSkipTableReader PKG=github.com/cockroachdb/cockroach/pkg/sql/distsqlrun TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1441595&tab=buildLog

```
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:316 +0x14a
  github.com/cockroachdb/cockroach/pkg/kv.GRPCTransportFactory()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/transport_race.go:102 +0x2c4
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendToReplicas()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1657 +0xbc
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendRPC()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:436 +0x319
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendSingleRange()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:518 +0x3fc
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendPartialBatch()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1441 +0x4a0
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).divideAndSendBatchToRanges()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1090 +0x1c30
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:732 +0x12f5
  github.com/cockroachdb/cockroach/pkg/internal/client.lookupRangeFwdScan()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/range_lookup.go:328 +0x490
  github.com/cockroachdb/cockroach/pkg/internal/client.RangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/range_lookup.go:198 +0xc2d
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).RangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:359 +0x11e
  github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).performRangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:449 +0x3d4
  github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).lookupRangeDescriptorInternal.func3()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:322 +0x166
  github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight.(*Group).doCall()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight/singleflight.go:128 +0x4c
==================
I190816 05:48:21.096500 5688 gossip/gossip.go:1531  [n1] node has connected to cluster via gossip
I190816 05:48:21.097015 5688 storage/stores.go:259  [n1] wrote 0 node addresses to persistent storage
W190816 05:48:21.098406 6565 storage/intentresolver/intent_resolver.go:818  failed to gc transaction record: could not GC completed transaction anchored at /Local/Range/Table/53/RangeDescriptor: node unavailable; try another peer
I190816 05:48:21.100885 5550 util/stop/stopper.go:542  quiescing; tasks left:
9      [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
1      [async] closedts-rangefeed-subscriber
I190816 05:48:21.101382 5550 util/stop/stopper.go:542  quiescing; tasks left:
8      [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
I190816 05:48:21.102056 5550 util/stop/stopper.go:542  quiescing; tasks left:
4      [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
I190816 05:48:21.102473 5550 util/stop/stopper.go:542  quiescing; tasks left:
2      [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
I190816 05:48:21.102785 5550 util/stop/stopper.go:542  quiescing; tasks left:
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
I190816 05:48:21.103386 5550 util/stop/stopper.go:542  quiescing; tasks left:
1      [async] transport racer
I190816 05:48:21.253164 5809 kv/transport_race.go:108  transport race promotion: ran 21 iterations on up to 1009 requests
    --- FAIL: TestIndexSkipTableReader/InterleavedChildReverse (0.02s)
        testing.go:809: race detected during execution of test
    testing.go:809: race detected during execution of test

```

SHA: https://github.com/cockroachdb/cockroach/commits/8ebcdac113118ae5fbcaddeecd269f59399aea8c

Parameters:
```
TAGS=
GOFLAGS=-race -parallel=2
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestIndexSkipTableReader PKG=github.com/cockroachdb/cockroach/pkg/sql/distsqlrun TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1444182&tab=buildLog

```
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:518 +0x3d7
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendPartialBatch()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1441 +0x4a0
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).divideAndSendBatchToRanges()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1090 +0x1c30
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:732 +0x12f5
  github.com/cockroachdb/cockroach/pkg/internal/client.lookupRangeFwdScan()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/range_lookup.go:328 +0x490
  github.com/cockroachdb/cockroach/pkg/internal/client.RangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/range_lookup.go:198 +0xc2d
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).RangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:359 +0x11e
  github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).performRangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:449 +0x3d4
  github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).lookupRangeDescriptorInternal.func3()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:322 +0x166
  github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight.(*Group).doCall()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight/singleflight.go:128 +0x4c
==================
W190819 05:39:40.834112 6770 storage/intentresolver/intent_resolver.go:818  failed to gc transaction record: could not GC completed transaction anchored at /Table/SystemConfigSpan/Start: node unavailable; try another peer
I190819 05:39:40.833986 5200 util/stop/stopper.go:542  quiescing; tasks left:
9      [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
1      [async] closedts-rangefeed-subscriber
I190819 05:39:40.835563 5200 util/stop/stopper.go:542  quiescing; tasks left:
6      [async] storage.IntentResolver: cleanup txn records
1      [async] transport racer
1      [async] intent_resolver_gc_batcher
1      [async] closedts-rangefeed-subscriber
I190819 05:39:40.837603 5200 util/stop/stopper.go:542  quiescing; tasks left:
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
I190819 05:39:40.837834 5200 util/stop/stopper.go:542  quiescing; tasks left:
1      [async] transport racer
I190819 05:39:40.925992 5663 storage/queue.go:518  [n1,s1] rate limited in MaybeAdd (gc): node unavailable; try another peer
I190819 05:39:40.926296 5663 storage/queue.go:518  [n1,s1] rate limited in MaybeAdd (merge): node unavailable; try another peer
I190819 05:39:40.926421 5663 storage/queue.go:518  [n1,s1] rate limited in MaybeAdd (split): node unavailable; try another peer
I190819 05:39:40.926511 5663 storage/queue.go:518  [n1,s1] rate limited in MaybeAdd (replicate): node unavailable; try another peer
I190819 05:39:40.926634 5663 storage/queue.go:518  [n1,s1] rate limited in MaybeAdd (replicaGC): node unavailable; try another peer
I190819 05:39:40.926729 5663 storage/queue.go:518  [n1,s1] rate limited in MaybeAdd (raftlog): node unavailable; try another peer
I190819 05:39:40.926827 5663 storage/queue.go:518  [n1,s1] rate limited in MaybeAdd (raftsnapshot): node unavailable; try another peer
I190819 05:39:40.926919 5663 storage/queue.go:518  [n1,s1] rate limited in MaybeAdd (consistencyChecker): node unavailable; try another peer
I190819 05:39:40.927051 5663 storage/queue.go:518  [n1,s1] rate limited in MaybeAdd (timeSeriesMaintenance): node unavailable; try another peer
I190819 05:39:40.970536 5628 kv/transport_race.go:108  transport race promotion: ran 24 iterations on up to 1008 requests
    --- FAIL: TestIndexSkipTableReader/InterleavedChildReverse (0.01s)
        testing.go:809: race detected during execution of test
    testing.go:809: race detected during execution of test

```

--- a/pkg/sql/distsqlrun/index_skip_table_reader.go
+++ b/pkg/sql/distsqlrun/index_skip_table_reader.go
@@ -35,9 +35,6 @@ type indexSkipTableReader struct {
 	// currentSpan maintains which span we are currently scanning.
 	currentSpan int
 
-	// scratchKey is scratch space to build prefix keys for following iterations.
-	scratchKey []byte
-
 	// keyPrefixLen holds the length of the prefix of the index
 	// that we are performing a distinct over.
 	keyPrefixLen int
@@ -111,7 +108,6 @@ func newIndexSkipTableReader(
 		return nil, err
 	}
 	t.indexLen = len(index.ColumnIDs)
-	t.scratchKey = make([]byte, 0, t.indexLen)
 
 	cols := immutDesc.Columns
 	if returnMutations {
@@ -185,17 +181,23 @@ func (t *indexSkipTableReader) Next() (sqlbase.EncDatumRow, *distsqlpb.ProducerM
 			}
 		}
 
+		var key []byte
 		{
 			// Fetch the PartialKey in a separate scope so that it cannot be reused
 			// outside this scope.
-			key, err := t.fetcher.PartialKey(t.keyPrefixLen)
+			unsafeKey, err := t.fetcher.PartialKey(t.keyPrefixLen)
 			if err != nil {
 				t.MoveToDraining(err)
 				return nil, &distsqlpb.ProducerMetadata{Err: err}
 			}
-			// Modifying key in place would corrupt our current row, so make a copy.
-			// Use an append here to ensure we copy all elements.
-			t.scratchKey = append(t.scratchKey[:0], key...)
+			// Modifying unsafeKey in place would corrupt our current row, so make a
+			// copy.
+			// A key needs to be allocated every time because it is illegal to mutate
+			// batch requests which would end up happening if we used scratch space.
+			// TODO(asubiotto): Another option would be to construct a key after
+			//  returning a row (at the top of this loop).
+			key = make([]byte, len(unsafeKey))
+			copy(key, unsafeKey)
 		}
 
 		row, _, _, err := t.fetcher.NextRow(t.Ctx)
@@ -220,15 +222,15 @@ func (t *indexSkipTableReader) Next() (sqlbase.EncDatumRow, *distsqlpb.ProducerM
 			//  encounter one. We also need a test to ensure that we behave properly
 			//  when we encounter maximal (i.e. all nulls) prefix keys.
 			for i := 0; i < (t.indexLen - t.keyPrefixLen + 1); i++ {
-				t.scratchKey = append(t.scratchKey, 0xff)
+				key = append(key, 0xff)
 			}
-			t.spans[t.currentSpan].Key = t.scratchKey
+			t.spans[t.currentSpan].Key = key
 		} else {
 			// In the case of reverse, this is much easier. The reverse fetcher
 			// returns the key retrieved, in this case the first key smaller
 			// than EndKey in the current span. Since EndKey is exclusive, we
 			// just set the retrieved key as EndKey for the next scan.
-			t.spans[t.currentSpan].EndKey = t.scratchKey
+			t.spans[t.currentSpan].EndKey = key
 		}
 
 		// If the changes we made turned our current span invalid, mark that

